### ISSUE TITLE
Topological Sorting

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#34343]
**Use case**

Backup and restore tables based on their dependencies.

**Describe the solution you'd like**

`ORDER BY x DEPENDS ON y`

Where `y` has the same type as `x` or Nullable of the same type, or it is an array of elements of the same type as `x`.

```
what         dependencies
'Moscow'     ['Russia']
'Russia'     ['Europe', 'Asia']
'Europe'     []
'Asia'       []
```

`SELECT what FROM places ORDER BY what DEPENDS ON dependencies`

will give

```
Europe
Asia
Russia
Moscow
```

Similarly, on:

```
what         dependencies
'Moscow'     'Russia'
'Russia'     'Europe'
'Europe'     ''
```

the same query will give

```
Europe
Russia
Moscow
```

To implement it, we can maintain the set of already processed items and a queue of postponed items.
If an item has dependency that has not been already processed, we will place it to the end of queue.
We also track, what was the last number of items in the queue after the item, when the item from the queue was pushed back to the end of queue again. If the number does not change, we understand it as a loop or a broken dependency and process item nevertheless.

PS. Not sure if the algorithm is correct. More info https://en.wikipedia.org/wiki/Topological_sorting

**Comment by bharath000** [comment]
@alexey-milovidov I am new to this project, Can I start with this Feature?

**Comment by alexey-milovidov** [comment]
@bharath000 Yes, it is possible.
You can start with https://clickhouse.com/docs/en/development/developer-instruction

**Comment by bharath000** [comment]
- If we are doing "Backup and restore tables based on their dependencies." What is the metric/relationship that determines dependencies between tables.
- Do we need to building any new query functionality with "DEPENDS" like shown "SELECT what FROM places ORDER BY what DEPENDS ON dependencies"? Or Is this an example?
- Correct me If we are building query like above than it should be table level (not column level).

**Comment by bharath000** [comment]
@alexey-milovidov Hi I am able to set-up and build the server and client, Can you help me with above doubts and also pointers to start with.

**Comment by bharath000** [comment]
@tavplubix Can you help with ^^^ above

**Comment by bharath000** [comment]
> 

Thanks you so much for the Info. Can you also confirm that in the AST of  **SELECT Query**, **DEPENDS** ON** is always child of **ORDER BY**

**Comment by tavplubix** [comment]
Yep

**Comment by Ilyua** [comment]
Hi! Can I ask to assign this InternTask for me?

**Comment by nvevg** [comment]
> Hi! Can I ask to assign this InternTask for me?
Oh, started working on this issue just a couple of hours ago :) Does task have to be assigned to someone to propose a PR (AFAIK, it doesn't, but I could be missing something)?

**Comment by Ilyua** [comment]
I found this task marked as available in InternTasks so I decided to ask to assign it) AFAIK no assignation is not required to propose PR. Although I think we can work in parallel :)

**Comment by Ilyua** [comment]
I saw that some tasks are assigned to workers here https://github.com/ClickHouse/ClickHouse/issues/58394 so I think we should follow this processes

**Comment by Ilyua** [comment]
@alexey-milovidov can you assign this task in a proper way?

**Comment by Ilyua** [comment]
@rschu1ze

**Comment by rschu1ze** [comment]
Assignment is not a requirement for working on this.

@Ilyua and @nvevg To avoid that we end up with two PRs/implementation for the same task, perhaps you like to collaborate on this one?

**Comment by Ilyua** [comment]
I am open to any discussions and collaboration. It is my first attempt to participate in open source project. So I would be pleased if we can discuss further steps with more experienced colleagues.

One more question about idea: we want add this feature only for one column of dependencies, don't we? In other words there couldn't be nested dependencies defining more complex graph. 

So after some time I created pipeline for myself: 

1.  I need to add new `ParserKeyword` "DEPENDS ON"
2. I need to set flag "depends_on_order_by" while parsing in `ParserSelectQuery.cpp`
3. I need to extend `SortStep` in which I will read DEPENDS ON column and make Topological sorting in `InterpreterSelectQuery.cpp`
4. I need to add new logic kind of "if depends on order by call sorting with new `SortStep` and `SortDescription` in `InterpreterSelectQuery.cpp` and that's all? 
5. Maybe I need to add new kind of `QueryNode`? I am trying to avoid work with inners of AST. 

So can you help me by critics of this pipeline? I want to make clear steps before implementation. 

My thoughts based on @rschu1ze previous  [work](https://github.com/ClickHouse/ClickHouse/pull/57875) and reading architecture of ClickHouse in ContirbutionGuide

Thank you!

**Comment by Ilyua** [comment]
@rschu1ze Can I make PR with some questions and minor additions inside code to discuss whether chosen path of mine is true or we need discuss questions firstly and only then make PR? Thank you!

**Comment by Ilyua** [comment]
Can anybody help me?

**Comment by zakr600** [comment]
I would like to take this project
